### ISSUE TITLE
Allow link_whole to work when no functions in static library are called

### ISSUE THREAD

**Issue by chergert** [mesonbuild/meson#2374]
In Builder, we're linking all of our plugins into a single shared library to reduce the number of modules that have to be opened at start-up. To keep our linking times fast, I think it would be useful if we could have the plugins split into their own `static_library()` and then only incur the final link cost when rebuilding.

However, since no function's are called in these static libraries, they're getting dropped despite using `link_whole: static_archive_plugins` (presumably from `-Wl,as-needed`).

It would be nice if meson could do something like:

```
-Wl,--push-state,--no-as-needed $link_options -Wl,--pop-state
```

So that nothing needs to be done to ensure that all the symbols from the `.a` are available.

I still want `-Wl,as-needed`, just not for the archives (which already have symbol visibility set). I'm not positive, but this might also help with situations where static constructors are used to register types/resources/etc.

**Comment by nirbheek** [comment]
This seems like really dumb behaviour for the linker to have. I'll test various linkers and see what happens.

**Comment by chergert** [comment]
@sarum9in my use case typically revolves around `GResources` that are linked into a `.a` continuing to work after linking into a `.so` or executable. They use static-constructors and those (at least in the past) would be dropped by the linker.

My other uses are registering dynamic `GObject` types from static-constructors, but that is the same problem as above.

**Comment by sarum9in** [comment]
@chergert I am not familiar with some of your terms. By static constructor do you mean self-registration code like [this](https://gist.github.com/sarum9in/58b7f32a9ba6346f0afe256dd733e001#file-static-cpp-L4)?

```cpp
static __attribute__((constructor)) void init() {
   // register itself in shared factory object
}
```

I believe static initialization I introduced is generating the same code, with the only difference that it is cross-platform C++ code that is not dependent on gcc-specific attributes.

I tested both
- linking static self-registering plugins into executable via `link_whole`
- linking static self-registering plugins into shared library via `link_whole` and linking shared library with executable

Both worked on gcc, so I wonder how is your use case different.

**Comment by chergert** [comment]
In Builder we use a vendor'd copy of gconstructor. https://gitlab.gnome.org/GNOME/gnome-builder/blob/master/src/libide/gconstructor.h

We had to do this https://gitlab.gnome.org/GNOME/gnome-builder/blob/master/src/plugins/gnome-builder-plugins.c to get things working a while back, or the `.so` would get dropped when checking with `ldd`.

**Comment by sarum9in** [comment]
@chergert hold on a second, maybe I misunderstood something from the start. The function you are calling is `gnome_builder_plugins_init`, and this function is located in the shared library. So your use case layout is this:
1. `exe` links with `shared_lib` and `exe` calls `gnome_builder_plugins_init` to make sure linker does not drop shared_lib, and this is the place your problem appeared, `shared_lib` was not getting linked with `exe`
1. `shared_lib` uses `link_whole` to compound all static plugin libraries, and this works just fine with `link_whole`

If my understanding is correct, you need `--no-as-needed` on the (1) link, between `exe` and `shared_lib`. Your workaround specifically ensures that linker will not drop (1), it has nothing to do with plugins linked into `shared_lib`. You should be able to verify that by running `meson -Db_asneeded=false` on your project with `gnome_builder_plugins_init` dropped entirely.

The proper fix is what I proposed in #3046, I hope to get this patched into 0.46 release. Idea is to allow link_whole for shared libraries as well and link them with `--no-as-needed` to make sure linker does not drop anything.

**Comment by chergert** [comment]
@sarum9in That sounds like a reasonable summation. That work was a while ago so it's a bit fuzzy in my head. What I think happened is I ran into a number of issues in different approaches and finally landed on something that worked.

I think my initial desire was that we'd have a `libgnome-builder-plugins.a` that got statically linked into the `gnome-builder` executable and each of the tests. But I couldn't get that to work, even with link whole due to gresources getting dropped. So then I went down the `.so` approach instead of a static library. That too failed, but for a different reason (the as-needed getting dropped). And at that point, I even tried overriding the flags specifically for the target, and it still failed.

At that point, and too much time burnt, I just gave up and added a symbol I could dummy call.
